### PR TITLE
Add basic sampling support

### PR DESCRIPTION
**Pull request by indrekj**
See README changes for more information.

This is very similar to jaeger-client-ruby.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,6 +31,16 @@ end
...
 
+### Samplers
+
+#### Const sampler
+
+`Const` sampler always makes the same decision for new traces depending on the initialization value. Set `sampler` to: `Zipkin::Samplers::Const.new(true)` to mark all new traces as sampled.
+
+#### Probabilistic sampler
+
+`Probabilistic` sampler samples traces with probability equal to `rate` (must be between 0.0 and 1.0). Set `sampler` to `Zipkin::Samplers::Probabilistic.new(rate: 0.1)` to mark 10% of new traces as sampled.
```
Comment (take-five) [inline README.md]: Isn't `probabilistic` sampler with sample rate eq `1.0` the same as `const` sampler? What do you think of having just a `sample_rate` configuration option?

Comment (indrekj) [reply README.md]: It is same. I saw it implemented like this in multiple places: e.g. https://github.com/jaegertracing/jaeger-client-go#sampling . Maybe it's just one calculation less, although it should be really fast.

Comment (take-five) [reply README.md]:
```
def sample?(*)
  return true if @rate == 1.0

  ...
end
```

this would cover this optimization

Comment (take-five) [reply README.md]: I'm fine with current implementation though, just thought that it'd be easier to understand and set up

File: lib/zipkin/samplers/probabilistic.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    class Probabilistic
+      def initialize(rate: 0.001)
+        if rate < 0.0 || rate > 1.0
+          raise "Sampling rate must be between 0.0 and 1.0, got #{rate.inspect}"
+        end
+        @boundary = TraceId::TRACE_ID_UPPER_BOUND * rate
+      end
+
+      def sample?(trace_id)
+        @boundary >= trace_id.to_i(16)
```
Comment (deiwin) [inline lib/zipkin/samplers/probabilistic.rb]: So trace IDs are already random hex numbers? Does this decision need to be made multiple times for the same trace? I.e. is it important that the decision is deterministic, given a certain trace?

File: lib/zipkin/samplers/probabilistic.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    # Sample a portion of traces using trace_id as the random decision
+    class Probabilistic
+      def initialize(rate: 0.001)
+        if rate < 0.0 || rate > 1.0
+          raise "Sampling rate must be between 0.0 and 1.0, got #{rate.inspect}"
+        end
+        @boundary = TraceId::TRACE_ID_UPPER_BOUND * rate
+      end
+
+      def sample?(trace_id)
```
Comment (deiwin) [inline lib/zipkin/samplers/probabilistic.rb]: Can a trace be accessed here via `trace_id`? If not, then the interface doesn't support smarter samplers that need to inspect the trace. I can imagine there's use for samplers that e.g. always allow traces that aren't similar to any other traces to get maximal coverage.

File: lib/zipkin/samplers/probabilistic.rb (hunk 2 above)
Comment (indrekj) [reply lib/zipkin/samplers/probabilistic.rb]: Yes, they are basically `rand(2**64)`. There's no need to do it multiple times. It's done once and then propagated using http headers or whatever the transport is. So I don't think it has to be deterministic but I don't think it hurts. I followed https://github.com/jaegertracing/jaeger-client-go/blob/master/sampler.go#L110

File: lib/zipkin/samplers/probabilistic.rb (hunk 3 above)
Comment (indrekj) [reply lib/zipkin/samplers/probabilistic.rb]: No it cannot. Sampling is decided before any of the spans are created. go injects two parameters `trace_id` and `operation name` https://github.com/jaegertracing/jaeger-client-go/blob/master/sampler.go#L43 It seems that what you're describing is similar to this https://github.com/jaegertracing/jaeger-client-go/blob/master/sampler.go#L198

Comment (deiwin) [reply lib/zipkin/samplers/probabilistic.rb]: Is it okay that you'll break compatibility with old samplers when you want to add `operatrion_name` as a parameter? If not, `trace_id` should be named variable.

Comment (indrekj) [reply lib/zipkin/samplers/probabilistic.rb]: changed to named argument

**Pull request merged**